src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php

```diff
@@ -45,18 +50,52 @@ public function __construct(string $type, array $options = [], bool $allowAdd =
     public static function getSubscribedEvents(): array
     {
         return [
-            FormEvents::PRE_SET_DATA => 'preSetData',
+            FormEvents::PRE_SET_DATA => 'preSetData', // deprecated
+            FormEvents::POST_SET_DATA => ['postSetData', 255], // as early as possible
```

Comment:
is it actually allowed to modify the structure of the form on postSetData ?

Reply:
It is, the form is not cloned and can be modified by reference https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Form/Form.php#L327.
The only event that forbids it is `POST_SUBMIT` thanks to this check https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Form/Form.php#L260, in such case `$this->defaultDataSet` is true, but not enough, only data cannot further be changed.

Reply:
>To me, this is too late as those child fields won't be populated.

They will be: https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Form/Form.php#L776.

Reply:
I've refactored the tests to ensure data is mapped consistently: https://github.com/symfony/symfony/pull/51041/commits/2662d0b09bfea6c50d84b2cb89c6a690e92ca85b.